### ISSUE TITLE
Weight sharing not consistent with paper

### ISSUE THREAD

**Issue by gshaikovski** [lucidrains/perceiver-pytorch#67]
Hi Phil,

Want to confirm the reason behind this design choice:
https://github.com/lucidrains/perceiver-pytorch/blob/c3d505a997a6e3521e83d7d2bf57cb8b62e3fbd6/perceiver_pytorch/perceiver_pytorch.py#L194-L210

In the paper, they say that they tie all the latent transformer weights. However in this implementation, TF in the first layer is not shared with the rest.

<img width="461" alt="image" src="https://github.com/lucidrains/perceiver-pytorch/assets/28565097/4806c865-b7b5-4f70-b11c-057ff90bcaab">

It should probably be
```python
            for block_ind in range(self_per_cross_attn):
                self_attns.append(nn.ModuleList([
                    get_latent_attn(_cache=True, key = block_ind),
                    get_latent_ff(_cache=True, key = block_ind)
                ]))
```

What do you think?

**Comment by JingrongZheng** [comment]
> Hi Phil,
> 
> Want to confirm the reason behind this design choice:
> 
> [perceiver-pytorch/perceiver_pytorch/perceiver_pytorch.py](https://github.com/lucidrains/perceiver-pytorch/blob/c3d505a997a6e3521e83d7d2bf57cb8b62e3fbd6/perceiver_pytorch/perceiver_pytorch.py#L194-L210)
> 
> Lines 194 to 210 in [c3d505a](/lucidrains/perceiver-pytorch/commit/c3d505a997a6e3521e83d7d2bf57cb8b62e3fbd6)
> 
>  for i in range(depth): 
>      should_cache = i > 0 and weight_tie_layers 
>      cache_args = {'_cache': should_cache} 
>   
>      self_attns = nn.ModuleList([]) 
>   
>      for block_ind in range(self_per_cross_attn): 
>          self_attns.append(nn.ModuleList([ 
>              get_latent_attn(**cache_args, key = block_ind), 
>              get_latent_ff(**cache_args, key = block_ind) 
>          ])) 
>   
>      self.layers.append(nn.ModuleList([ 
>          get_cross_attn(**cache_args), 
>          get_cross_ff(**cache_args), 
>          self_attns 
>      ])) 
> In the paper, they say that they tie all the latent transformer weights. However in this implementation, TF in the first layer is not shared with the rest.
> 
> <img alt="image" width="461" src="https://private-user-images.githubusercontent.com/28565097/274903289-4806c865-b7b5-4f70-b11c-057ff90bcaab.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjE0ODg5NjIsIm5iZiI6MTc2MTQ4ODY2MiwicGF0aCI6Ii8yODU2NTA5Ny8yNzQ5MDMyODktNDgwNmM4NjUtYjdiNS00ZjcwLWIxMWMtMDU3ZmY5MGJjYWFiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDI2VDE0MjQyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMwMmUyNjMwM2ViYjdmMjc4NzY4OWJjNDI3N2FhYjMwNTRlMTc5MzkxYTQwNGMwOTIzMThiMTFiZGZmNDMwYmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.qUG_B9HmkExDp7tZFVlLAGHZWvoMsaPKU5t_Lhb2p6Y">
> It should probably be
> 
>             for block_ind in range(self_per_cross_attn):
>                 self_attns.append(nn.ModuleList([
>                     get_latent_attn(_cache=True, key = block_ind),
>                     get_latent_ff(_cache=True, key = block_ind)
>                 ]))
> What do you think?

I agree with you, all latent transformers should shared the same weight according to the paper. I met the same problem and found there's a mistake in this code.
